distributed/worker_state_machine.py

```diff
@@ -1885,11 +1885,13 @@ def _transition_generic_released(
         self._purge_state(ts)
         recs: Recs = {}
         for dependency in ts.dependencies:
+            dependency.dependents.discard(ts)
             if (
                 not dependency.waiters
                 and dependency.state not in READY | PROCESSING | {"memory"}
             ):
                 recs[dependency] = "released"
+        ts.dependencies.clear()
```

Comment:
I do not think we should do this. A rather big guiding principle for the WSM was the notion of transitioning tasks back to released when changing from a "runnable" to a "dependency" or the other way round. In todays states, this would mean a transition from ready/executing to fetch/flight, etc. Effectively the released transition is always used when we need a clean slate but it's never used to actually forget the task or change otherwise static task attributes. By modifying dependencies and dependents here you are effectively altering the graph.

Part of the "transition over released state" logic is discussed in https://github.com/dask/distributed/issues/4413 and other tickets. Im not saying that this has to be this way forever but this is how the thing was built. Changing this is non-trivial. Besides implementing a bunch of additional, otherwise entirely redundant, transition we'd possibly also expose ourselves to more edge cases.